apps/sales/views.py

```diff
@@ -1457,6 +1457,428 @@ def put(self, request, pk, *args, **kwargs):
+        custom_data["sale_receipt"] = new_sale_receipt_data
+        sale_receipt_id = new_sale_receipt_data.get("sale_receipt_id", None)  # Fetch sale_receipt_id from mew instance
+        logger.info('SaleReceipt - created*')
+
+        #create History for Lead with assignemnt date as current and leave the end date as null.
+        update_fields = {'sale_receipt_id':sale_receipt_id}
+
+        return build_response(1, "Record created successfully", custom_data, status.HTTP_201_CREATED)
+    
+    def put(self, request, *args, **kwargs):
+        return self.update(request, *args, **kwargs)
+    
+    @transaction.atomic
+    def update(self, request, pk, *args, **kwargs):
+
+        # ----------------------------------- D A T A  V A L I D A T I O N -----------------------------#
+        """
+        All the data in request will be validated here. it will handle the following errors:
+        - Invalid data types
+        - Invalid foreign keys
+        - nulls in required fields
+        """
+        # Get the given data from request
+        given_data = request.data
+        errors = {}
+
+        # Validate SaleReceipt Data
+        sale_receipt_data = given_data.pop('sale_receipt', None)  # parent_data
+        if sale_receipt_data:
+            sale_receipt_data['sale_receipt_id'] = pk
+            sale_receipt_error = validate_multiple_data(self, [sale_receipt_data], SaleReceiptSerializer, [])
+            if sale_receipt_error:
+                errors["sale_receipt"] = sale_receipt_error
+
+        if errors:
+            return build_response(0, "ValidationError :", errors, status.HTTP_400_BAD_REQUEST)
+
+        # ------------------------------ D A T A   U P D A T I O N -----------------------------------------#
+        custom_data = {
+            'sale_receipt':{}
+            }
+        
+        '''
+        This "save_assignments_history_with_end_date" function saves the instance in 'LeadAssignmentHistory' model
+        by updating the end date.
+
+        def save_assignments_history_with_end_date(self):
+            now = datetime.now()
+            end_date = now.strftime('%Y-%m-%d %H:%M:%S')
+            history_data_set = LeadAssignmentHistory.objects.filter(sale_receipt_id=pk).order_by('created_at').last()
+            history_data_set.end_date = end_date # update with end date
+            history_data_set.save() # save the record
+            logger.info(f'last history_id : {history_data_set}')
+
+
+        This "create_new_history" function creates the new instance in 'LeadAssignmentHistory' model
+        with required fields to be updated.
+
+        def create_new_history(self,update_fields):
+            assignment_history = {}
+            assignments_history = generic_data_creation(self, [assignment_history], LeadAssignmentHistorySerializer, update_fields)
+            assignments_history = assignments_history if assignments_history else []
+            logger.info('LeadAssignmentHistory - created*') '''
```

Comment:
What is above commented? It is not specific to sale_receipt

Reply:
yeah Thulasi removed , I pasted sale_reciept code which was sent by Pramod , he mistakenly added it Thulasi.